cxx/distributions/bigram.hh

```diff
@@ -0,0 +1,88 @@
+        assert(c >= ' ' && c <= '~');
+    }
+    size_t char_to_index(const char c) const {
+        assert_valid_char(c);
+        return c - ' ';
+    }
+    char index_to_char(const size_t i) const {
+        const char c = i + ' ';
+        assert_valid_char(c);
+        return c;
+    }
+    std::vector<size_t> string_to_indices(const std::string& str) const {
+        // Convert the string to a vector of indices between 0 and `num_chars`,
+        // with a start/stop symbol at the beginning/end.
+        std::vector<size_t> inds = {num_chars};
+        for (const char& c : str) {
+            inds.push_back(char_to_index(c));
+        }
+        inds.push_back(num_chars);
+        return inds;
+    }
+public:
+    double alpha = 1;  // hyperparameter for all transition distributions.
+    size_t num_chars = '~' - ' ' + 1;  // printable ASCII without DEL.
+    std::vector<DirichletCategorical> transition_dists;
+    PRNG *prng;
+
+    Bigram(PRNG *prng) {
+        this->prng = prng;
+        const size_t total_chars = num_chars + 1;  // Include a start/stop symbol.
+
+        // The distribution at index `i` represents `p(x_{i+1} | x_i)`.
+        transition_dists.reserve(total_chars);
+        for (size_t i = 0; i != total_chars; ++i) {
+            transition_dists.emplace_back(prng, total_chars);
+        }
+    }
+    void incorporate(const std::string& x) {
+        const std::vector<size_t> indices = string_to_indices(x);
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            transition_dists[indices[i]].incorporate(indices[i + 1]);
+        }
+    }
+    void unincorporate(const std::string& s) {
+        const std::vector<size_t> indices = string_to_indices(s);
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            transition_dists[indices[i]].unincorporate(indices[i + 1]);
+        }
+    }
+    double logp(const std::string& s) const {
+        const std::vector<size_t> indices = string_to_indices(s);
+        double total_logp = 0.0;
+        for (size_t i = 0; i != indices.size() - 1; ++i) {
+            total_logp += transition_dists[indices[i]].logp(indices[i + 1]);
+        }
+        return total_logp;
+    }
+    double logp_score() const {
+        return std::transform_reduce(
+            transition_dists.cbegin(), 
+            transition_dists.cend(), 
+            0, 
+            std::plus{},
+            [&](auto d) -> double {return d.logp_score(); }
+        );
+    }
+    std::string sample() {
+        std::string sampled_string;
+        // Sample the first character conditioned on the stop/start symbol.
+        size_t current_ind = transition_dists[num_chars].sample();
+        // Sample additional characters until the stop/start symbol is sampled.
+        while (current_ind != num_chars) {
+            sampled_string += index_to_char(current_ind);
+            current_ind = transition_dists[current_ind].sample();
+        }
```

Comment:
Technically we run into the same issue here as in `logp`, where previously sampled letters should change the distribution of future letters. After each sample, you could `incorporate` the observed transition into the underlying `transition_dist`, and then at the end, call `unincorporate` on the generated string.

Reply:
Fixed.